airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/RecordDiffer.kt

```diff
@@ -0,0 +1,264 @@
+        val diffs = matches.filter { it.isMismatch() }
+        return if (diffs.isEmpty()) {
+            null
+        } else {
+            diffs.joinToString("\n") { it.prettyPrintMismatch() }
+        }
+    }
+
+    private inner class MatchingRecords(
+        val expectedRecord: OutputRecord?,
+        val actualRecord: OutputRecord?,
+    ) {
+        fun isMismatch(): Boolean =
+            (expectedRecord == null && actualRecord != null) ||
+                (expectedRecord != null && actualRecord == null) ||
+                !recordsMatch(expectedRecord, actualRecord)
+
+        fun prettyPrintMismatch(): String {
+            return if (expectedRecord == null) {
+                "Unexpected record (${generateRecordIdentifier(actualRecord!!)}): $actualRecord"
+            } else if (actualRecord == null) {
+                "Missing record (${generateRecordIdentifier(expectedRecord)}): $expectedRecord"
+            } else {
+                "Incorrect record (${generateRecordIdentifier(actualRecord)}):\n" +
+                    generateDiffString(expectedRecord, actualRecord).prependIndent("  ")
+            }
+        }
+
+        private fun recordsMatch(
+            expectedRecord: OutputRecord?,
+            actualRecord: OutputRecord?,
+        ): Boolean =
+            (expectedRecord == null && actualRecord == null) ||
+                (expectedRecord != null &&
+                    actualRecord != null &&
+                    generateDiffString(expectedRecord, actualRecord).isEmpty())
+
+        private fun generateRecordIdentifier(record: OutputRecord): String {
+            // If the PK is an empty list, then don't include it
+            val pk: List<AirbyteValue> = extractPrimaryKey(record)
+            val pkString = if (pk.isEmpty()) "" else "pk=$pk"
+
+            if (cursor != null) {
+                val cursor: AirbyteValue = extractCursor(record)
+                return "$pkString, cursor=$cursor"
+            } else {
+                return pkString
+            }
+        }
+
+        private fun generateDiffString(
+            expectedRecord: OutputRecord,
+            actualRecord: OutputRecord,
+        ): String {
+            val diff: StringBuilder = StringBuilder()
+            // Intentionally don't diff loadedAt / rawId, since those are generated dynamically by
+            // the destination.
+            if (expectedRecord.extractedAt != actualRecord.extractedAt) {
+                diff.append(
+                    "extractedAt: Expected ${expectedRecord.extractedAt}, got ${actualRecord.extractedAt}\n"
+                )
+            }
+            if (expectedRecord.generationId != actualRecord.generationId) {
+                diff.append(
+                    "generationId: Expected ${expectedRecord.generationId}, got ${actualRecord.generationId}\n"
+                )
+            }
+            if (expectedRecord.airbyteMeta != actualRecord.airbyteMeta) {
```

Comment:
there's a bug here that I need to fix - airbyteMeta is a JsonNode, which has issues with comparison (`IntNode(42) != LongNode(42)`)

so I need to either make it a real struct, or write something to compare jsonnodes for equality